src/layer/x86/gridsample_x86.cpp

```diff
@@ -0,0 +1,456 @@
+// Tencent is pleased to support the open source community by making ncnn available.
+//
+// Copyright (C) 2023 THL A29 Limited, a Tencent company. All rights reserved.
+//
+// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// https://opensource.org/licenses/BSD-3-Clause
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "gridsample_x86.h"
+
+#if __SSE2__
+#include <emmintrin.h>
+#include "sse_mathfun.h"
+#if __AVX__
+#include <immintrin.h>
+#include "avx_mathfun.h"
+#if __AVX512F__
+#include "avx512_mathfun.h"
+#endif // __AVX512F__
+#endif // __AVX__
+#endif // __SSE2__
+#include "x86_usability.h"
+#include "cpu.h"
+
+namespace ncnn {
+
+#include "gridsample_compute_blob.h"
+#include "gridsample_apply_interpolation.h"
+
+GridSample_x86::GridSample_x86()
+{
+#if __SSE2__
+    support_packing = true;
+#endif // __SSE2__
+}
+
+int GridSample_x86::forward(const std::vector<Mat>& bottom_blobs, std::vector<Mat>& top_blobs, const Option& opt) const
+{
+    const Mat& bottom_blob = bottom_blobs[0];
+    const Mat& grid = bottom_blobs[1];
+    Mat& top_blob = top_blobs[0];
+    int elempack = bottom_blob.elempack;
+
+    int channels = bottom_blob.c;
+    int dims = bottom_blob.dims;
+    size_t elemsize = bottom_blob.elemsize;
+
+    int outw, outh, outd;
+    Mat offset_blob, value_blob;
+
+    Mat grid_p1;
+    if (grid.elempack != 1)
+    {
+        convert_packing(grid, grid_p1, 1, opt);
+    }
+    else
+    {
+        grid_p1 = grid;
+    }
+
+    if (dims == 3)
+    {
+        outw = permute_fusion == 0 ? grid_p1.h : grid_p1.w;
+        outh = permute_fusion == 0 ? grid_p1.c : grid_p1.h;
+
+        top_blob.create(outw, outh, channels, elemsize, elempack, opt.blob_allocator);
+        if (top_blob.empty())
+            return -100;
+
+        if (sample_type == GridSample::Interpolation_BILINEAR)
+        {
+            offset_blob.create(outw, outh, 4, elemsize, 1, opt.workspace_allocator);
+            value_blob.create(outw, outh, 2, elemsize, 1, opt.workspace_allocator);
```

Comment:
packed layout can be more efficient, less pointer access usually means less tlb miss

```cpp
offset_blob.create(outw, outh, elemsize * 4, 4, opt.workspace_allocator);
value_blob.create(outw, outh, elemsize * 2, 2, opt.workspace_allocator);
```

or even merge the offset and value blob into one
```cpp
offset_value_blob.create(outw, outh, elemsize * 6, 6, opt.workspace_allocator);
```


The same applies for bicubic and 4d routines